jopa-impl/src/main/java/cz/cvut/kbss/jopa/sessions/validator/CardinalityConstraintsValidator.java

```diff
@@ -57,17 +58,22 @@ public void validate(Object identifier, Field field, Object value) {
 
     @Override
     public void validate(Object identifier, FieldSpecification<?, ?> attribute, Object attributeValue) {
-        if (!(attribute instanceof Attribute)) {
-            // Only proper attributes can have cardinality constraints
-            return;
-        }
-        final Attribute<?, ?> att = (Attribute<?, ?>) attribute;
-        final int valueCount = extractValueCount(attributeValue);
-        for (ParticipationConstraint pc : att.getConstraints()) {
-            validateParticipationConstraint(identifier, att.getJavaField(), valueCount, pc);
-        }
-        if (att.getConstraints().length == 0) {
-            validateNonEmpty(identifier, att, valueCount);
+        // Only proper attributes can have cardinality constraints
+        if (attribute instanceof Attribute) {
+            final Attribute<?, ?> att = (Attribute<?, ?>) attribute;
+            final int valueCount = extractValueCount(attributeValue);
+            for (ParticipationConstraint pc : att.getConstraints()) {
+                validateParticipationConstraint(identifier, att.getJavaField(), valueCount, pc);
+            }
+            if (att.getConstraints().length == 0) {
+                validateNonEmpty(identifier, att, valueCount);
+            }
+        } else if (attribute instanceof QueryAttribute) {
+            final QueryAttribute<?, ?> queryAtt = (QueryAttribute<?, ?>) attribute;
+            final int valueCount = extractValueCount(attributeValue);
+            for (ParticipationConstraint pc : queryAtt.getConstraints()) {
+                validateParticipationConstraint(identifier, queryAtt.getJavaField(), valueCount, pc);
+            }
```

Comment:
Try to extract common code from these two `if`s into a separate method.

Reply:
This is problematic, because after passing the attribute to the new method as the common ancestor `FieldSpecification`, I loose the information that I got from the `if` in the first place and the `getConstraints()` method is only available from the inheriting interfaces. I've made it at least a little nicer by putting the repeating line `final int valueCount = extractValueCount(attributeValue);` before the `if`s.

Reply:
Actually, I've just found a way to do it, I've commited it, but the PR is already merged, should I reopen it, or leave it like this?

Reply:
Try committing again and open a new PR.